docs: graphql api page and other incremental improvements (#17516)
gatsbyjs/gatsby: docs/docs/graphql-api.md

---
title: GraphQL API
---

import { GraphqlApiQuery } from "../../www/src/components/api-reference/doc-static-queries"
import APIReference from "../../www/src/components/api-reference"

A great advantage of Gatsby is a built-in data layer that combines data from any and all data sources you link into. Data is collected at [build time](/docs/glossary#build) and automatically assembled into a [schema](/docs/glossary#schema) that defines how data can be queried throughout components and pages in your site.

This doc serves as a reference for GraphQL features built into Gatsby, including methods for querying and sourcing data, and customizing GraphQL for your site's needs.

## Getting Started with GraphQL

GraphQL doesn't need to be installed in a new Gatsby site, it is already included. A schema is inferred and created when you run `gatsby develop` or `gatsby build` and can be [explored](/docs/running-queries-with-graphiql/) at `http://localhost:8000/___graphql` when a Gatsby site successfully compiles.

## Sourcing Data

> Data needs to be [sourced](/docs/content-and-data/) -- or added to the GraphQL schema -- to be queried and pulled into pages using GraphQL. Gatsby uses [source plugins](/plugins/?=gatsby-source) to pull in, or source, data.

Reviewer: Nitpick about correct em-dashes :3

```suggestion
Data needs to be [sourced](/docs/content-and-data/) — or added to the GraphQL schema — to be queried and pulled into pages using GraphQL. Gatsby uses [source plugins](/plugins/?=gatsby-source) to pull in, or source, data.
```

**Note**: GraphQL isn't necessary, you can still [use Gatsby without GraphQL](/docs/using-gatsby-without-graphql/)

Sourcing data with an existing plugin requires installing necessary packages and adding the plugin to the plugins array in the `gatsby-config` with any necessary configurations. To [source data from the filesystem](/docs/sourcing-from-the-filesystem/):

> ```sh

Reviewer: ```suggestion
```shell
```

npm install --save gatsby-source-filesystem
```

```js:title=gatsby-config.js
module.exports = {
  siteMetadata: {
    title: `Your Site Name`,
  },
  plugins: [
    // highlight-start
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `src`,
        path: `${__dirname}/src/`,
      },
    },
    // highlight-end
  ],
}
```

> You can also [create custom plugins](https://www.gatsbyjs.org/docs/creating-plugins/) to fit your own use cases and pull in data however you want.

Reviewer: ```suggestion
You can also [create custom plugins](/docs/creating-plugins/) to fit your own use cases and pull in data however you want.
```

## Querying for Data

> Data can be queried inside pages, components, or the `gatsby-node.js` file, using one of these options:

Reviewer: I'd also like to have a section or at least a note that you can't mix page queries and static queries in the same file. You also can't have multiple page queries or static queries in one file.

---

> title: GraphQL API

Reviewer 2: ```suggestion
title: GraphQL API
tableOfContentsDepth: 2
```
The `params` and `returns` are pretty repetitive, so I'd recommend suppressing those. 2 seems to be the magic number when I run this locally.

This doc serves as a reference for GraphQL features built into Gatsby, including methods for querying and sourcing data, and customizing GraphQL for your site's needs.

## Getting Started with GraphQL

GraphQL doesn't need to be installed in a new Gatsby site, it is already included. A schema is inferred and created when you run `gatsby develop` or `gatsby build` and can be [explored](/docs/running-queries-with-graphiql/) at `http://localhost:8000/___graphql` when a Gatsby site successfully compiles.

## Sourcing Data

Data needs to be [sourced](/docs/content-and-data/) — or added to the GraphQL schema — to be queried and pulled into pages using GraphQL. Gatsby uses [source plugins](/plugins/?=gatsby-source) to pull in, or source, data.

**Note**: GraphQL isn't necessary, you can still [use Gatsby without GraphQL](/docs/using-gatsby-without-graphql/)

Sourcing data with an existing plugin requires installing necessary packages and adding the plugin to the plugins array in the `gatsby-config` with any necessary configurations. To [source data from the filesystem](/docs/sourcing-from-the-filesystem/):

> ```shell

Reviewer 2: I think we should remove these install and configuration steps for the API doc and link off to that content instead. See my suggestion above on "Sourcing data"

npm install --save gatsby-source-filesystem
```

```js:title=gatsby-config.js
module.exports = {
  siteMetadata: {
    title: `Your Site Name`,
  },
  plugins: [
    // highlight-start
    {
      resolve: `gatsby-source-filesystem`,
      options: {
        name: `src`,
        path: `${__dirname}/src/`,
      },
    },
    // highlight-end
  ],
}
```

You can also [create custom plugins](/docs/creating-plugins/) to fit your own use cases and pull in data however you want.

> ## Querying for Data

Reviewer 2: ```suggestion
## Query components and hooks
```
"Querying for data" is mentioned throughout the docs for so many different purposes that it would help to be more specific. For this API doc, headlining the _thing_ developers are looking for might clear things up a bit.

Data can be queried inside pages, components, or the `gatsby-node.js` file, using one of these options:

> - `pageQuery`

Reviewer 2: ```suggestion
- The `pageQuery` component
```

> - `StaticQuery`

Reviewer 2: ```suggestion
- The `StaticQuery` component
```

> - `useStaticQuery`

Reviewer 2: ```suggestion
- The `useStaticQuery` hook
```

> **Note**: Because of how Gatsby processes GraphQL queries, you can't mix page queries and static queries in the same file. You also can't have multiple page queries or static queries in one file.

Reviewer 2: ```suggestion
**Note**: Because of how Gatsby processes GraphQL queries, you can't mix page queries and static queries in the same file. You also can't have multiple page queries or static queries in one file.

For information on page and non-page components as they relate to queries, check out the docs guide on [building with components](/docs/building-with-components/#how-does-gatsby-use-react-components)
```

### `pageQuery`

> You can have one page query per page. A page query can take GraphQL arguments for variables in your queries. A [page is made in Gatsby](/docs/page-creation/) by any React component in the `src/pages` folder, or by calling the `createPage` action and using a component in the `createPage` options -- meaning a `pageQuery` won't work in any component, only components that meet this criteria.

Reviewer 2: ```suggestion
`pageQuery` is a built-in component that retrieves information from the data layer in Gatsby pages. You can have one page query per page. It can take GraphQL arguments for variables in your queries.

A [page is made in Gatsby](/docs/page-creation/) by any React component in the `src/pages` folder, or by calling the `createPage` action and using a component in the `createPage` options -- meaning a `pageQuery` won't work in any component, only components that meet this criteria.
```

Also, refer to the [guide on querying data in pages with page query](/docs/page-query/).

#### Params

A page query isn't a method, but is assigned a `graphql` string with a valid query inside it as an exported constant:

```javascript
export const pageQuery = graphql`
  query HomePageQuery {
    site {
      siteMetadata {
        description
      }
    }
  }
`
```

**Note**: the `const` that gets exported doesn't need to be called `pageQuery`, it can have anything, Gatsby just looks for an exported `graphql` string from the file.

#### Returns

A page query returns a data object passed into the props of the page component of the file the query is in.

```javascript
// highlight-start
const HomePage = ({ data }) => {
  // highlight-end
  return (
    <div>
      Hello!
      {data.site.siteMetadata.description} // highlight-line
    </div>
  )
}
```

### `StaticQuery`

> Gatsby provides a `StaticQuery` component that can be used in any component to get data from GraphQL. You can only have one `StaticQuery` per page, in order to include the data you need from multiple sources you can use one query with multiple root fields. It cannot take variables as arguments.

Reviewer 2: ```suggestion
StaticQuery is a built-in component for retrieving data from Gatsby’s data layer in non-page components, such as a header, navigation, or any other child component.

You can only have one `StaticQuery` per page: in order to include the data you need from multiple sources, you can use one query with multiple [root fields](). It cannot take variables as arguments.
```
Is there something we can link to about root fields? or add an explanation?

Author: I like [this blog post from Apollo](https://blog.apollographql.com/the-anatomy-of-a-graphql-query-6dffa9e9e747) that explains a bunch of those concepts. I don't see anything we have that explains root fields, maybe I add it to the GraphQL concepts page with some other explanations sort of like that blog post includes?

Author: I added this image and an explanation in the GraphQL concepts page that this can link to for now:

![image](https://user-images.githubusercontent.com/21114044/65079757-17715500-d95d-11e9-9583-5341b569c965.png)

I also added the above blog post to the Further Reading section since that's what I referenced in making a Gatsby specific, simpler example.

Also, refer to the [guide on querying data in components with static query](/docs/static-query/).

#### Params

The `StaticQuery` component takes two values as props:

- query: a `graphql` query string
- render: a component with access to the data returned

```jsx
<StaticQuery
  query={graphql`
    query HeadingQuery {
      site {
        siteMetadata {
          title
        }
      }
    }
  `}
  render={data => (
    <header>
      <h1>{data.site.siteMetadata.title}</h1>
    </header>
  )}
/>
```

#### Returns

> StaticQuery returns data in a render prop:

Reviewer 2: ```suggestion
The StaticQuery component returns `data` in a `render` prop:
```

    </header>
  )}
  // highlight-end
/>
```

### `useStaticQuery`

The `useStaticQuery` hook can be used similar to `StaticQuery` in any component or page, but doesn't require the use of a component and render prop.

Because it is a React hook, the [rules of hooks](https://reactjs.org/docs/hooks-rules.html) apply and you'll need to be using React and ReactDOM version 16.8.0 or later. Because of how queries currently work in Gatsby, only one instance of `useStaticQuery` is supported in each file.

Also, refer to the [guide on querying data in components with useStaticQuery](/docs/use-static-query/).

#### Params

The `useStaticQuery` hook takes one argument:

- query: a `graphql` query string

```jsx
const data = useStaticQuery(graphql`
  query HeaderQuery {
    site {
      siteMetadata {
        title
      }
    }
  }
`)
```

#### Returns

The `useStaticQuery` hook returns data in an object:

```jsx
const data = useStaticQuery(graphql`
  query HeaderQuery {
    site {
      siteMetadata {
        title
      }
    }
  }
`)
return (
  // highlight-start
  <header>
    <h1>{data.site.siteMetadata.title}</h1>
  </header>
  // highlight-end
)
```

> ## GraphQL Fragments

Reviewer 2: ```suggestion
## Query fragments
```
Adjusting to be consistent with the above headline change for `Query components and hooks`

For understanding the parts of a query refer to the [conceptual guide](/docs/graphql-concepts/#understanding-the-parts-of-a-query).

### GraphQL query arguments

GraphQL queries can take arguments to alter how the data is returned. The logic for these arguments is handled internally by Gatsby. Arguments can be passed in to fields at any level of the query.

Different nodes can take different arguments based off of the nature of the node.

The arguments you can pass to collections (like arrays or long lists of data - ex. `allFile`, or `allMdx`) are:

- [`filter`](/docs/graphql-reference#filter)
- [`limit`](/docs/graphql-reference#limit)
- [`sort`](/docs/graphql-reference#sort)
- [`skip`](/docs/graphql-reference#skip)

The arguments you can pass to a `date` field are:

- [`formatString`](/docs/graphql-reference#dates)
- [`locale`](/docs/graphql-reference#dates)

The arguments you can pass to an `excerpt` field are:

- [`pruneLength`](/docs/graphql-reference#excerpt)
- [`truncate`](/docs/graphql-reference#excerpt)

### Graphql query operations

Other built-in configurations can be used in queries

- [`Alias`](/docs/graphql-reference#alias)
- [`Group`](/docs/graphql-reference#group)

> 

Reviewer 2: ```suggestion

For examples, refer to the [Query recipes](/docs/recipes/#6-querying-data) and [GraphQL Query options Reference Guide](/docs/graphql-reference/).
```
This URL will change slightly when we break the recipes up into pages, but we'll do a global find and replace to make sure there aren't broken links.
